// @algorandfoundation/tealscript/dist/bin/tealscript.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs = __importStar(require("fs"));
const argparse_1 = require("argparse");
const typescript_1 = __importDefault(require("typescript"));
const ts_morph_1 = require("ts-morph");
const glob_1 = require("glob");
const compiler_1 = __importDefault(require("../lib/compiler"));
require("dotenv/config");
const version_1 = require("../version");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function processFile(filename, parsed) {
    // Output dir for artifacts
    const dir = parsed.output;
    const tsConfigFilePath = path_1.default.join(process.cwd(), typescript_1.default.findConfigFile(filename, typescript_1.default.sys.fileExists, 'tsconfig.json'));
    const project = new ts_morph_1.Project({
        tsConfigFilePath,
    });
    if (project.getCompilerOptions().experimentalDecorators !== true) {
        // eslint-disable-next-line no-console
        console.warn(`WARNING: 'experimentalDecorators' is not enabled in your tsconfig.json. TEALScript is manually overriding this option (in memory) for compilation. This will slightly increase the TEALSCript compile time. Please set 'experimentalDecorators' to 'true' in ${tsConfigFilePath}.`);
        project.compilerOptions.set({ experimentalDecorators: true });
    }
    const options = {
        algodPort: process.env.ALGOD_PORT ? Number(process.env.ALGOD_PORT) : undefined,
        algodServer: process.env.ALGOD_SERVER,
        algodToken: process.env.ALGOD_TOKEN,
        disableWarnings: parsed.disable_warnings,
        disableOverflowChecks: parsed.unsafe_disable_overflow_checks,
        disableTypeScript: parsed.unsafe_disable_typescript,
        srcPath: filename,
        project,
        cwd: process.cwd(),
        skipAlgod: parsed.skip_algod,
    };
    const compilers = compiler_1.default.compileAll(options);
    compilers.forEach(async (compilerPromise) => {
        const compiler = await compilerPromise;
        const { name } = compiler;
        const approvalPath = path_1.default.join(dir, `${name}.approval.teal`);
        const clearPath = path_1.default.join(dir, `${name}.clear.teal`);
        const abiPath = path_1.default.join(dir, `${name}.arc4.json`);
        const appPath = path_1.default.join(dir, `${name}.arc32.json`);
        const srcmapPath = path_1.default.join(dir, `${name}.src_map.json`);
        const lsigPath = path_1.default.join(dir, `${name}.lsig.teal`);
        const arc56Path = path_1.default.join(dir, `${name}.arc56.json`);
        if (fs.existsSync(approvalPath))
            fs.rmSync(approvalPath);
        if (fs.existsSync(clearPath))
            fs.rmSync(clearPath);
        if (fs.existsSync(abiPath))
            fs.rmSync(abiPath);
        if (fs.existsSync(appPath))
            fs.rmSync(appPath);
        if (fs.existsSync(srcmapPath))
            fs.rmSync(srcmapPath);
        if (fs.existsSync(lsigPath))
            fs.rmSync(lsigPath);
        if (fs.existsSync(arc56Path))
            fs.rmSync(arc56Path);
        if (compiler.teal.lsig.length > 0) {
            fs.writeFileSync(lsigPath, compiler.teal.lsig.map((t) => t.teal).join('\n'));
            fs.writeFileSync(srcmapPath, JSON.stringify(compiler.sourceInfo, null, 2));
            return;
        }
        fs.writeFileSync(approvalPath, compiler.teal.approval.map((t) => t.teal).join('\n'));
        fs.writeFileSync(clearPath, compiler.teal.clear.map((t) => t.teal).join('\n'));
        fs.writeFileSync(abiPath, JSON.stringify(compiler.arc4Description(), null, 2));
        fs.writeFileSync(srcmapPath, JSON.stringify(compiler.sourceInfo, null, 2));
        fs.writeFileSync(appPath, JSON.stringify(compiler.arc32Description(), null, 2));
        fs.writeFileSync(arc56Path, JSON.stringify(compiler.arc56Description(), null, 2));
    });
}
const parser = new argparse_1.ArgumentParser({
    description: 'Argparse example',
});
parser.add_argument('input', { help: 'TEALScript source file(s) to compile', nargs: '*' });
parser.add_argument('output', { help: 'output directory for artifacts' });
parser.add_argument('-v', '--version', { action: 'version', version: version_1.VERSION });
parser.add_argument('--disable-warnings', {
    help: 'disables compiler warnings',
    action: 'store_true',
    default: false,
});
parser.add_argument('--unsafe-disable-overflow-checks', {
    help: 'disables overflow checks for ABI uint/ufixed types',
    action: 'store_true',
    default: false,
});
parser.add_argument('--unsafe-disable-typescript', {
    help: 'disables typescript diagnostics',
    action: 'store_true',
    default: false,
});
parser.add_argument('--skip-algod', {
    help: 'Skip algod compilation. This results in no source mapping and no algod assebmler error checking. The emitted TEAL may not properly compile without additional steps.',
    action: 'store_true',
    default: false,
});
const parsed = parser.parse_args();
parsed.input.forEach((f) => (0, glob_1.globSync)(f).forEach((gf) => {
    processFile(gf, parsed);
}));
